alerter.cpp

```diff
@@ -1,32 +1,64 @@
-#include <iostream>
-#include <assert.h>
+#include "alerter.h"
 
-int alertFailureCount = 0;
+int gloabl::alertFailureCount = 0;
 
-int networkAlertStub(float celcius) {
-    std::cout << "ALERT: Temperature is " << celcius << " celcius.\n";
+int TestAlerter::networkAlertStub(const float celcius) {
+    std::cout << "TEST ALERT: Temperature is " << celcius << " celcius.\n";
     // Return 200 for ok
     // Return 500 for not-ok
     // stub always succeeds and returns 200
     return 200;
 }
 
-void alertInCelcius(float farenheit) {
-    float celcius = (farenheit - 32) * 5 / 9;
+int ProductionAlerter::networkAlertStub(const float celcius) {
+    std::cout << "PRD ALERT: Temperature is " << celcius << " celcius.\n";
+
+    if(MAX_THRESHOLD > celcius) {
+        return 200;
+    }
+    else {
+        return 500;
+    }
+}
+
+float Alerter::convertFarenheitToCelcius(const float farenheit) {
+    return (farenheit - 32) * 5 / 9;
+}
+
+void Alerter::addFailure() {
+    gloabl::alertFailureCount += 1;
+}
+
+void Alerter::validateFailureCount() {
+    assert (gloabl::alertFailureCount > 0);
+}
+
+void Alerter::alertInCelcius(const float farenheit) {
+    float celcius = convertFarenheitToCelcius(farenheit);
     int returnCode = networkAlertStub(celcius);
+
     if (returnCode != 200) {
         // non-ok response is not an error! Issues happen in life!
         // let us keep a count of failures to report
         // However, this code doesn't count failures!
         // Add a test below to catch this bug. Alter the stub above, if needed.
-        alertFailureCount += 0;
+        addFailure();
+        validateFailureCount();
     }
 }
 
 int main() {
-    alertInCelcius(400.5);
-    alertInCelcius(303.6);
-    std::cout << alertFailureCount << " alerts failed.\n";
-    std::cout << "All is well (maybe!)\n";
+    //Run alerter in test environement
+    Alerter *alerterObj = new TestAlerter();
+
+    //Run alerter in production environement 
+    //Alerter *alerterObj = new ProductionAlerter();
+ 
+    alerterObj->alertInCelcius(400.5);
+    alerterObj->alertInCelcius(303.6);
+    alerterObj->alertInCelcius(303.8);
+
+    std::cout << gloabl::alertFailureCount << " alerts failed.\n";

```

Comment:
How we can separate assert from production code if keeping in main ?

Reply:
Usually production code is developed as a seperate library. Then it is consumed in the test project for testing.

But since we are only starting here, we should first try keeping production code in separate files and test code in separate files. Here the problem is that production code is also present in main.

If main had contained only test code and the test code calls only APIs of production code, then goal is achieved.

Hope this is clear. Let me know if there are doubts 👍